### ISSUE TITLE
Add memory size estimation for `ParquetMetadata`

### ISSUE THREAD

**Issue by alamb** [apache/arrow-rs#1729]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
In https://github.com/influxdata/influxdb_iox, caching is very important to performance. Part of what is cached in memory is a `ParquetMetadata` structure. In order to effectively cache this data (and free it when under memory pressure) we need to know an accurate estimate of the heap it is using. 

**Describe the solution you'd like**
I would like a function such as the following that accurately estimates the memory usage of parquet metadata:

```rust
impl ParquetMetadata {
    ...

    /// In-memory size in bytes, including `self`.
    pub fn size(&self) -> usize {
        // recursively 
    }

   ...
}
```

**Describe alternatives you've considered**
I believe @domodwyer is considering caching only the parts that we need (rather than the `ParquetMetadata` object itself) in which case we would likely not need this feature in IOx. I think it would still be generally helpful though

**Additional context**
See https://github.com/influxdata/influxdb_iox/pull/4661 for example of usecase

**Comment by alamb** [comment]
I believe @crepererum  may be working on this

**Comment by crepererum** [comment]
I think I might not. After a quick look at the data structures, the memory estimation is quite a mess, especially for the parquet schema data.

**Comment by alamb** [comment]
I am going to try and find time to do this sometime this week

**Comment by Xuanwo** [comment]
> In https://github.com/influxdata/influxdb_iox, caching is very important to performance. Part of what is cached in memory is a `ParquetMetadata` structure.

Seems a typo here, https://github.com/influxdata/influxdb_iox is empty

**Comment by Xuanwo** [comment]
Oh, I see. This issue created in 2022, and `influxdb_iox` has been renamed to `influxdb`. What a history.

**Comment by alamb** [comment]
> Oh, I see. This issue created in 2022, and `influxdb_iox` has been renamed to `influxdb`. What a history!

Yeah, the actual history is more tortured -- if you care it is here https://www.influxdata.com/blog/the-plan-for-influxdb-3-0-open-source/

TLDR is that we made the influxdb_iox repo private

**Comment by alamb** [comment]
Here is a proposed PR with a fix: https://github.com/apache/arrow-rs/pull/5965
